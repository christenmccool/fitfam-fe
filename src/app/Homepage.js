import React, {useState, useEffect, useContext} from 'react';
import { Link as RouterLink, useSearchParams } from 'react-router-dom';
import moment from 'moment';

import Container from '@mui/material/Container';
import DatePicker from '@mui/lab/DatePicker';
import Box from '@mui/material/Box';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';

import UserContext from '../auth/UserContext';
import PostingList from '../postings/PostingList';
import WorkoutList from '../workouts/WorkoutList';


/** FitFam homepage
 * 
 * Anonymous user:
 * Shows WorkoutList for today's date
 * WorkoutList -> WorkoutCardList -> WorkoutCard 
 * Links to WorkoutSearchPage
 * 
 * Logged in user:
 * Shows PostingList for today's date
 * PostingList -> PostingCardList -> PostingCard 
 * 
 * Routed at /
 */
const Homepage = () => {
  const { user } = useContext(UserContext);
  const [searchParams, setSearchParams] = useSearchParams("");

  const initialDate = searchParams.get('date') || moment().format("YYYY-MM-DD");
  const [date, setDate] = useState(initialDate);

  useEffect(() => {
    const newDate = searchParams.get('date');
    if (newDate) {
      setDate(newDate);
    } else {
      setSearchParams({date: moment().format("YYYY-MM-DD")});
    }
  }, [searchParams, setSearchParams])
  
  return (
    <Container maxWidth="md" align="center">
      <Box mt={4} sx={{display:"flex", justifyContent:"center"}}>
        <DatePicker
          label="Date"
          value={date}
          onChange={(newDate) => setSearchParams({date: moment(newDate).format("YYYY-MM-DD")})}
          renderInput={(params) => <TextField {...params} sx={{backgroundColor: "#FFF", input: {fontSize:'20px', p:1}}}/>}
        />
      </Box>
      <Box mb={4}>
        {user ?
            <PostingList date={date}/>
          :
          <Box>
            <WorkoutList date={date} />
            <Button 
              component={RouterLink}
              to="/workouts"
              variant="contained" 
              size="large"
              sx={{ mt: 3 }}
            >
              <Typography variant="h6">Search for more workouts</Typography>
            </Button>
          </Box>
        }
      </Box>
    </Container>
  )
}

export default Homepage;