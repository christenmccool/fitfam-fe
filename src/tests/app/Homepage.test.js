import React from "react";
import { screen, render, act} from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import AdapterMoment from '@mui/lab/AdapterMoment';
import LocalizationProvider from '@mui/lab/LocalizationProvider';

import Homepage from "../../app/Homepage";

import FitFamApi from '../../api/api';
jest.mock('../../api/api');

beforeEach(() => {
  FitFamApi.getFeaturedWorkouts.mockResolvedValue([
      {
        id: 1, 
        name:"Test Workout Name 1", 
        description:"Do this hard thing", 
        category: "featured", 
        scoreType: "Reps", 
        createDate: "20220322",
        featuredDate: "20220322",
        createBy: null
      },
      {
        id: 2, 
        name:"Test Workout Name 2", 
        description:"Do this other hard thing", 
        category: "featured", 
        scoreType: "Reps", 
        createDate: "20220322",
        featuredDate: "20220322",
        createBy: null
      }
  ])
})

it('renders without crashing', async function () {
  await act(async () => {
    render(
      <MemoryRouter>
        <LocalizationProvider dateAdapter={AdapterMoment}>
          <Homepage />
        </LocalizationProvider>
      </MemoryRouter>
    );
  });
});


it("matches snapshot", async function () {
  await act(async () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LocalizationProvider dateAdapter={AdapterMoment}>
          <Homepage />
        </LocalizationProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});


it("renders list of workouts and search button", async function () {
  render(
    <MemoryRouter>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <Homepage />
      </LocalizationProvider>
    </MemoryRouter>
  );
  expect(await screen.findByText("Featured Workout", {exact: false})).toBeInTheDocument();
  expect(screen.getByText("Test Workout Name 1")).toBeInTheDocument();
  expect(screen.getByText("Test Workout Name 2")).toBeInTheDocument();
  expect(screen.getByText( "Search for more workouts")).toBeInTheDocument();

  screen.debug();
});





