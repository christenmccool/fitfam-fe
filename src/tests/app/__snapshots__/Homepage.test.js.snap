// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    align="center"
    class="MuiContainer-root MuiContainer-maxWidthSm css-ead4d9-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-19rm58x"
    >
      <div
        class="MuiBox-root css-1l4w6pd"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1tazx64-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
          >
            Date
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              aria-label="Choose date, selected date is Mar 25, 2022"
              aria-readonly="true"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              type="text"
              value="03/25/2022"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1g8cwq9"
              >
                <span>
                  Date
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        align="center"
        class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-19rm58x"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Loading
          </h6>
        </div>
      </div>
      <a
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButtonBase-root  css-1tgvzpz-MuiButtonBase-root-MuiButton-root"
        href="/workouts"
        tabindex="0"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        >
          Search for more workouts
        </h6>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
